### ISSUE TITLE
Proposal for new architecture of ContinuousFactor class

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#782]
![class](https://cloud.githubusercontent.com/assets/1461453/21132332/ae2df6a2-c113-11e6-9cac-4a7a20151df9.png)

This structure will help us in keeping all the different distributions separate and the user will be able to use all the different distributions from ContinuousFactor only.

**Comment by khalibartan** [comment]
@ankurankan I guess according to UML arrows should be in opposite direction :P, considering base class is being inherited in gaussian, drichlet, custom distribution class. Seems okay to me.

**Comment by ankurankan** [comment]
@khalibartan Haha.. This was my first attempt at UML Diagrams. Didn't know it's supposed to be like that :P. I will fix that.

**Comment by yashu-seth** [comment]
@ankurankan If the users have to provide an instance of the distribution to ```ContinuousFactor``` then why not use the individual distributions directly? If a particular method is unavailable in that distribution it will fall back to its parent class' (```ContinuousFactor```) implementation.

**Comment by ankurankan** [comment]
@yashu-seth No, the users won't have to provide instances. They could just pass a string for the name of the distribution along with the parameters for the distribution and the `ContinuousFactor` would create the object accordingly. So the API would look something like this:

```python
# Creating using the name of the distribution
phi = ContinuousFactor(['A', 'B'], 'multivariate_gaussian', mu=[1, 2], cov=[[0.1, 0.3], [0.3, 0.2]])
# Creating using a custom distribution function
phi = ContinuousFactor(['A', 'B'], pdf_fun)
# Creating using an instance of child classes of BaseDistribution.
phi = ContinuousFactor(['A', 'B'], multivariate_gaussian_instance)
```
What do you think?

**Comment by khalibartan** [comment]
@ankurankan I think for this kind of implementation we can look into scikit-learn for a reference. I agree with this kind of API, it is convenient this way.

**Comment by yashu-seth** [comment]
@ankurankan This seems convenient. But it would require a number of if-else conditions in ```ContinuousFactor```. 

@khalibartan Which module uses a similar implementation in scikit-learn?

**Comment by khalibartan** [comment]
@yashu-seth All the modules that use kernel in back uses a similar kind of an API (I meant similar not same). Like SVM's etc. I think to some extend if-else conditions can be avoided using dictonaires key being the string and values being definition of appropriate pdf function corresponding to that string.

**Comment by ankurankan** [comment]
@yashu-seth I don't think it will require a lot of if else. The only place where we will need an if-else would be the `__init__` method of `ContinuousFactor` for deciding which distribution to initialize but this can also be handled easily using a dict. For all the other methods we will just be calling `self.distribution.some_method`.

**Comment by yashu-seth** [comment]
@ankurankan @khalibartan Cool. With this we can allow only ```ContinuousFactor``` to be associated with pgmpy.models for continuous cases.

**Comment by kris-singh** [comment]
@ankurankan previously i had raised a question regarding what should be passed to assignment function a list of *args. Since the assignment function calls pdf which in case of scipy.stats distribution expects a string and in case of the present implementation of the continuous factor it accepts *args. Does this make sense to you guys?
I will start working on the given UML

**Comment by ankurankan** [comment]
@kris-singh I didn't get your first question. Can you please elaborate a bit? 

Regarding working on this issue, I think this will require quite significant changes and I would really not suggest you to start contributing with this issue :D. But if you would still like to go ahead with this, I would like to split this into smaller parts or I can write the base framework and you can build up on that. Please let me know.

**Comment by kris-singh** [comment]
@ankurankan  Yes sure i would like to contribute to this. So lets go ahead with the your plan.

**Comment by ankurankan** [comment]
@kris-singh Cool. I will create a new branch and put some basic structure and then you can take over it from there.

**Comment by yashu-seth** [comment]
@ankurankan I think we should begin work on this. A lot of other things depend on this.

**Comment by raghavg7796** [comment]
@ankurankan I would also like to work in the given UML.

**Comment by yashu-seth** [comment]
@ankurankan I think we can rename the current ContinuousFactor class to BaseDistribution, such that it acts as a parent class for classes like JointGaussianDistribution, CanonicalFactor etc. Then have a separate class named ContinuousFactor in which users can provide instances, or strings etc. If you want I can create the base structure. Then I will look into #774.

**Comment by ankurankan** [comment]
@yashu-seth @raghavg7796 I have been really busy with school. But I will have time from tomorrow so I will do this the first thing. And hopefully will have a PR ready for this by tomorrow.

**Comment by yashu-seth** [comment]
@ankurankan Any update on this? I am free these days so if you want I can take this up after discussing the exact details.

**Comment by ankurankan** [comment]
@yashu-seth Sorry for the delay. Yes please go ahead with this if you have time.

**Comment by yashu-seth** [comment]
> I think we can rename the current ContinuousFactor class to BaseDistribution, such that it acts as a parent class for classes like JointGaussianDistribution, CanonicalFactor etc. Then have a separate class named ContinuousFactor in which users can provide instances, or strings etc. 
>
@ankurankan What do you think?

**Comment by ankurankan** [comment]
@yashu-seth I got some time today and have started working on this. I am done with the Base class and the CustomDistribution class. Will open a PR in couple of minutes :).

**Comment by yashu-seth** [comment]
@ankurankan Okay :)

**Comment by csbuja** [comment]
Hey guys. I've been continuing this implementation -- we need the normalize function for continuous factor for inference. It's used in VariableElimination. Currently this method is not shown.

**Comment by csbuja** [comment]
Should we be building a class between BaseDistribution called BaseContinuousDistribution? I feel like having BaseDistribution calling integrate may be misleading. We assume that it must be continuous, without saying so.

**Comment by ankurankan** [comment]
@csbuja I didn't get exactly what you mean. Can you please elaborate a bit please ?

**Comment by csbuja** [comment]
In the pull request for this code, @yashu-seth asks you to move the marginalize member function, amongst others, to BaseDistribution, and you agree, since you want distributions where there aren't tricks in taking the product (e.g. not a gaussian) to fall back on those methods.

The method of implementing marginalization depends on if the distribution is continuous or discrete. Thus, we should add a class between {custom distribution, Gaussian, Dirichlet} and {BaseDistribution} called ContinuousDistribution, which takes advantage of the continuous properties of the random variable(s).

**Comment by michaelrempel** [comment]
My thought is that continuous variables offer a lot of possible solutions. Efficient code will favor a plugin style. The first attempt could be a robust but simple statistical examination with a cluster of linear approximations providing the solution as a continuous function approximation. For each X(i) in the solution space the associated vector would be used for all x values such that x>=X(i) and x<X(i+1) To extend this into infinite space X(0) might have it's value be ignored, and only the associated function considered. Each function would of course be in the form y=Ax+B. And a convention of A=NaN would return NaN. Thus when X(0) has A=NaN the solution has a lower bound in X(1). X(n) with A=NaN would also provide an upper bound. The only time I can see a small issue is with F(x) blowing up to infinity, where you want a particular discrete value to return NaN. In practice I think an exclusion range around the function is going to be out of practical bounds for almost every case.
I think a linear function is preferable to a x=f(x) value pair set with implied vector interpolation, because there is no way to extend the range to infinite approximation. I also prefer uneven bin sizes to optimize range functions. A binary tree representation might also provide quicker search. It need not be balanced and might be optimized for larger bins over smaller ones. It might also be optimized to return ordered lists of x values more quickly.

**Comment by ankurankan** [comment]
@michaelrempel Yes, I agree with your ideas but these are issues related to inference methods. Right now need to implement the base layer for continuous variables and then build these feature on that.

**Comment by khalibartan** [comment]
@ankurankan @yashu-seth Updates on this?

**Comment by ankurankan** [comment]
@khalibartan I think this is almost done but still needs a bit of cleanup. I am working on that.

**Comment by Erotemic** [comment]
Any update here, or places where help is needed?

**Comment by ankurankan** [comment]
A lot has changed since our last discussion on this. Some really nice probabilistic programming tools are now available to work with complex distributions (like pymc4, edward) which majorly use sampling methods and variational inference. With these tools now available, I am not sure where exactly does Bayesian Networks fit in the scene because, from a purely computational perspective, they are just factorized distributions. And since all these tools use functional representation, the factorized distributions can be very easily represented as a multilevel model. I can only think of two minor benefits of using Bayesian Networks instead of these probabilistic tools:
1. Can be computationally efficient: Using d-separation, we can figure out which parts of the model are not going to have any effect on the queried variable and not do any computation on those parts. Although, I am not sure how much improvement will this bring in practical cases.
2. Exact Inference: For closed-form distributions, we can use efficient Bayesian Network algorithms to compute exact distributions. Again, I am not sure how accurate are the approximate methods in such cases (I expect them to be quite accurate).

Based on these points, I am not sure if implementing continuous variables in pgmpy adds much value. Something that I think could possibly be useful is to implement an optimization layer between pgmpy and pymc4, which can reduce the model size based on the model structure and then use pymc4 to do the inference. Thoughts?

**Comment by Erotemic** [comment]
So I'm very much a beginner when it comes to probabilistic graphical models (computer-vision is my main topic area). I'm somewhat aware of the fact that computationally Bayes nets are equivalent to factorized distributions, but I wouldn't be able to explain the details. 

Something that pgmpy brings to the table for me is that I find its interface to be very intuitive and it gives me the ability to play with the distributions I'm working with. I took a quick look at edward and pymc4, and it seems like they have a higher knowledge prerequisite than pgmpy does.  Also pymc4 and edward are both tensorflow based, and given the option, I'm much more of a torch person. I highly value being able to interactively program my code. 

Given that you probably understand pymc4 / edward much better than I do, do you think their interface is on-par with how pgmpy lets you specify your conditional distributions and then combine them into a networkx-based graphical model?  (I also really like how pgmpy uses networkx). If not, then I think it makes sense to build some translation layer between pgmpy and some gpu-based probabilistic programming library. One thing about pgmpy I've noticed is that it can be a bit slow during inference.

**Comment by ankurankan** [comment]
@Erotemic Thanks for sharing your thoughts on this.

I see probabilistic programming (PP) tools (pymc4 and edward) to be broader than Bayes nets. Bayes nets are a specific way of specifying a distribution by factorizing it into smaller parts and the factorization can give us more insights into the distribution (like dependence structures, Causal inference). PP, on the other hand, allows us to represent all possible distributions and uses a functional form (something like Structural Equation Models (SEM) ). For example, a network (A ---> B) assuming both to be Gaussians can be specified as:
```
A = Normal(mean0, std0)
B = 0.5 * A + Normal(mean1, std1) # Equivalent to the conditional dist: P(B) = N((0.5 A + mean1); std1)
```
So, this functional form of representation allows specifying arbitrary distributions and theoretically network structures can be extracted from this form as well (path diagrams in SEM terminology). Though, you can't learn the network structure from the data in PP.

In inference terms, the main benefit of Bayes nets over the general representation is that it can be computationally more efficient because we exploit the dependence structure between the variables which is not usually done in PP. The other benefit is the ability to do exact inference and the computation is more transparent (or user has more control). PP usually uses Variational Inference (approximate the distribution with a simpler parameterized distribution and minimize the difference between them) or Sampling methods. 

In my opinion, the best way forward for pgmpy is to focus more on the features that heavily utilize/benefit from the network structure (structure learning, causal inference, exact inference) and for approximate inference, add an intermediate layer to convert pgmpy models to one of PP models and use them for inference.
